keps/sig-cli/5295-kyaml/README.md

```diff
@@ -0,0 +1,1457 @@
+
+**Important note:**
+This is EXCLUSIVELY about client-side formatting.
+This KEP *does not* propose any changes to apiserver
+or other components besides kubectl.
+
+Syntax has always been a magnet for arguments in our industry.
+We expect this to be no exception.
+Reasonable people will disagree with our choices,
+which is fine, but we should hold a strong opinion on this.
+
+YAML is a very large and complex specification.
+We expect that we will find places where KYAML may be less readable than conventional YAML.
+We should revisit this specification when that happens.
+Givent that KYAML is YAML it should be possible to revise
+without breaking anyone.
+
+### Risks and Mitigations
+
+Risk: Our encoder produces some output which is not valid YAML.
+Mitigation: That’s a bug. It’s hard to imagine a case that can’t be fixed.
+
+Risk: The result could be net worse than plain YAML.
+Mitigation: We will produce many examples and include them in the review process. We will solicit feedback from various groups on the result.
+
+Risk: This gives a false sense of safety, but doesn’t SOLVE  the problems with YAML, just avoids them.
+Mitigation: Docs?
+
+Risk: Like all APIs, people will come to depend on the details, which may prevent us from changing them.
+Mitigation: Be as precise as possible where we are confident, but leave room where we are not.
+
+## Design Details
+
+Creating an encoder for this format is not very technically challenging,
+but there are many details to consider.
+
+### Packaging
+
+<<[UNRESOLVED where it lives]>>
+
+The simplest thing to do is to write a new implementation of
+k8s.io/cli-runtime/pkg/printers.ResourcePrinter (in
+staging/src/k8s.io/cli-runtime/pkg/printers). However, this isn’t the most
+obviously reusable place for it.
+
+A more ecosystem-friendly solution is to write a new function or package under
+sigs.k8s.io/yaml](sigs.k8s.io/yaml) which can be used by other tools.
+
+<<[/UNRESOLVED]>>
+
+### Tooling
+
+k/k has a CI verifier (hack/verify-yamlfmt.sh) which looks for YAML files and
+ensures they are formatted “correctly”. The only meaning of “correct” is “block
+style”, and it corrupts KYAML input today (aside: it also does not handle
+multi-documents).  If we intend to convert examples, we need to fix this.
+
+We propose to add flags to it which specify to use “conventional” YAML or KYAML
+output, and set the default to KYAML (or to change the parent script to always
+specify KYAML).
+
+This tooling will preserve comments as best it can. Unfortunately, go-yaml (our
+YAML library) [does not always handle comments properly](https://github.com/yaml/go-yaml/issues/8).
+It is possible that some comments will be formatted wrongly, or lost entirely.
+The affected comments are not representable in YAML block-style, so this is not
+exactly a regression.
+
+A guiding principle for the tooling is that whatever output the tool produces,
+when fed back into the tool, should not be modified again.
```

Comment:
nit: figure out how to make this true while also not allowing people to capture KYAML fixtures and use them in tests that will break if we later tweak KYAML rendering